IPs/RSKIP106.md

```diff
@@ -0,0 +1,135 @@
+
+```
+toBase58Check('0x0d3bf5f30dda7584645546079318e97f0e1d044f', 111) => 'mgivuh9jErcGdRr81cJ3A7YfgbJV7WNyZV'
+```
+
+### deriveExtendedPublicKey
+
+The method `deriveExtendedPublicKey(string xpub, string path) returns (string)` takes as input an extended public key encoded in base58Check (as described in ref #4), an HD derivation path - a string of the form
+
+```
+S ::= n | n/S
+```
+
+with `n` an unsigned integer - also as described in #4 - and produces a base58Check encoded string corresponding to the extended public key obtained by deriving the given extended public key with the given path.
+
+#### Validations
+
+- `xpub` must be a valid base58check-encoded extended public key (either for mainnet or testnet).
+- `path` must be a valid path according to the BNF definition given above.
+
+#### Sample usage
+
+Given the base58Check-encoded extended public key `tpubD6NzVbkrYhZ4YHQqwWz3Tm1ESZ9AidobeyLG4mEezB6hN8gFFWrcjczyF77Lw3HEs6Rjd2R11BEJ8Y9ptfxx9DFknkdujp58mFMx9H5dc1r`, and the derivation path `2/3/4`, we would have that:
+
+```
+deriveExtendedPublicKey('tpubD6NzVbkrYhZ4YHQqwWz3Tm1ESZ9AidobeyLG4mEezB6hN8gFFWrcjczyF77Lw3HEs6Rjd2R11BEJ8Y9ptfxx9DFknkdujp58mFMx9H5dc1r', '2/3/4') => 'tpubDCGMkPKredy7oh6zw8f4ExWFdTgQCrAHToF1ytny3gbVy9GkUNK2Nqh7NbKbh8dkd5VtjUiLJPkbEkeg29NVHwxYwzHJFt9SazGLZrrU4Y4'
+```
+
+### extractPublicKeyFromExtendedPublicKey
+
+The method `extractPublicKeyFromExtendedPublicKey(string xpub) returns (bytes)` takes as input an extended public key encoded in base58Check (as described in ref #4) and returns a byte array corresponding to the compressed public key of the given extended public key.
+
+#### Validations
+
+- `xpub` must be a valid base58check-encoded extended public key (either for mainnet or testnet).
+
+#### Sample usage
+
+Given the base58Check-encoded extended public key `tpubDCwK6XsmwUx641qZ6Uyb2pcZXMCoFcyNBFZLbxwJT4iwgmUhY9DsW9FgaFUNE5s1KtM1upxpDuCvx1TbAqQMLMGFF5b1F6KpRpuHcLwcp4k` we would have that:
+
+```
+extractPublicKeyFromExtendedPublicKey('tpubDCwK6XsmwUx641qZ6Uyb2pcZXMCoFcyNBFZLbxwJT4iwgmUhY9DsW9FgaFUNE5s1KtM1upxpDuCvx1TbAqQMLMGFF5b1F6KpRpuHcLwcp4k') => '0x0360a7239100410a4f6019778fce803c6b168a9b1922a0eafbdff11e866070d9f8'
+```
+
+### getMultisigScriptHash
+
+The method `getMultisigScriptHash(uint8 minimumSignatures, bytes[] publicKeys) returns (bytes)` takes as input a minimum required number of signatures and an array of (either compressed or uncompressed) secp256k1 public keys and returns the 20-byte hash (i.e., output of `hash160`) of the Bitcoin N-of-M multisig scriptPub corresponding to the given arguments. It is important to notice that the compressed version of the given public keys will be used to generate the script. Therefore, public key compression will not make a difference in the function output.
```

Comment:
How are you planning to inject the "nonce" to generate different addresses for each contributor? Is this done by generating new pubkeys with increasing sequential numbers for BIP32?

Reply:
Exactly. Essentially each public key within `publicKeys` is itself a derived public key from a given master key (or another derived key, without loss of generality). That would be the case if implementing, for example, [BIP45](https://github.com/bitcoin/bips/blob/master/bip-0045.mediawiki). But the thing here is we are leaving the freedom of choice to the actual BTO implementer.